CMakeLists.txt

```diff
@@ -304,10 +306,10 @@ endif()
 # Set code coverage compiler flags
 if(WITH_CODE_COVERAGE)
     if(CMAKE_C_COMPILER_ID MATCHES "Clang")
-        set(CMAKE_C_FLAGS "-O0 ${CMAKE_C_FLAGS} -coverage")
+        set(CMAKE_C_FLAGS "-O1 ${CMAKE_C_FLAGS} -coverage")
     elseif(__GNUC__)
         # Some versions of GCC don't support -coverage shorthand
-        set(CMAKE_C_FLAGS "-O0 ${CMAKE_C_FLAGS} -ftest-coverage -fprofile-arcs -fprofile-values")
+        set(CMAKE_C_FLAGS "-O1 ${CMAKE_C_FLAGS} -ftest-coverage -fprofile-arcs -fprofile-values")
```

Comment:
Why -O1 instead of -O0?
This was changed to -O0 because the coverage data was lower quality with the -O1 we used previously, with lines being combined etc.

Reply:
Ref: https://gcc.gnu.org/onlinedocs/gcc-4.8.4/gcc/Gcov-and-Optimization.html

Reply:
I mentioned it in our group chat:

I am also getting this warning. I think it should be changed back to -O1:

 warning: the object size sanitizer has no effect at -O0, but is explicitly enabled: -fsanitize=array-bounds,bool,bounds,builtin,enum,float-divide-by-zero,function,integer-divide-by-zero,null,nonnull-attribute,object-size,pointer-overflow,return,returns-nonnull-attribute,shift,shift-base,shift-exponent,signed-integer-overflow,undefined,unsigned-integer-overflow,vla-bound,vptr,leak [-Winvalid-command-line-argument]

Reply:
Or perhaps we should separate coverage and sanitation?
It is quite annoying when the coverage visualization is obviously wrong to the human eye because of the optimizer combining things, and those are only the obvious instances. Makes it hard to trust it when using it for analyzing actual  code usage across platforms.

Reply:
I agree with @Dead2 that coverage and sanitation should be isolated to avoid conflicting requirements... I'm just worried we end up with a lot more CI runs (again)...

Reply:
Yeah, likely double the number of runs.

Reply:
Do we need to actually run the tests for the sanitizer runs, or can those be left out?

An alternative would be that each run does two compiles, one for testing/coverage and one for sanitizers. That cuts down on the number of CI runs, and with it the long queue and setup-times.

Reply:
Most of the time is spent in the unit tests, very little in the start up from what I can tell. The unit tests need to be run for both sanitizer and coverage.

Reply:
Queue time does not seem to get accounted anywhere, but it is often significant, not sure how much can be saved.
On Ubuntu, it seems that setup takes at least 13seconds, but does not include provisioning and "boot", not sure whether it includes git checkout.
So doing each after each other would save an undeterminate amount of time.

A workaround for the sanitizer problem could perhaps be to disable the "object-size" checker, and rather add that to non-O0 test runs.

Reply:
I thought doing two compiles in same CI run messes up caching compiler output...It would also require cleaning up build tree when doing build inside source directory, or wiping the separate build directory...

Reply:
@Dead2 are you suggesting only adding _object-size_ checker when `-O0` not found in `C_FLAGS`? That would be a simple solution.

Reply:
@mtl1979 Doing two compiles should not be a problem for comiler caches..  I was thinking the sanitizer could be built in a separate directory, thus also removing the need for testing out-of-sourcedir builds separately.

@nmoinvaz I guess that would work, yes.

Reply:
Ok it is done.